JSON logs do not hydrate template placeholders

### Discussed in https://github.com/verdaccio/verdaccio/discussions/2286

<div type='discussions-op-text'>

<sup>Originally posted by **capps1994** June 11, 2021</sup>
We have updated our Verdaccio instance, to 5.1.0 and now logging seems to be not working as expected.

In our logs we see `{"level":25,"time":1623421647904,"pid":8,"hostname":"e643fb945f2c","request":{"method":"GET","url":"/-/verdaccio/packages"},"user":null,"remoteIP":"172.18.0.1","status":304,"bytes":{"in":0,"out":0},"msg":"@{status}, user: @{user}(@{remoteIP}), req: '@{request.method} @{request.url}', bytes: @{bytes.in}/@{bytes.out}"}`

We have the following in our config `logs:`
  `- { type: stdout, format: pretty, level: http }`

Has anything changed?
  </div>
<br>


@juanpicado I was in the process of opening a PR against `master` (for Verdaccio 6) but it looks like the unit tests for `packages/logger` are busted as are a number of other unit tests for other packages. I read the [Contributing Guidelines](https://github.com/verdaccio/verdaccio/blob/master/CONTRIBUTING.md) and followed the instructions for [_Running test_](https://github.com/verdaccio/verdaccio/blob/master/CONTRIBUTING.md#running-test) but continue (even after some troubleshooting) to receive errors:
```text
packages/logger test$ cross-env NODE_ENV=test BABEL_ENV=test jest --runTestsByPath
│ FAIL test/logger.spec.ts
│   ● Test suite failed to run
│     Cannot find module '@verdaccio/core' from 'test/logger.spec.ts'
│     > 1 | import { warningUtils } from '@verdaccio/core';
│         | ^
│       2 |
│       3 | import { logger, setup } from '../src';
│       4 |
│       at Resolver.resolveModule (../../node_modules/.pnpm/jest-resolve@27.3.1/node_modules/jest-resolve/build/resolver.js:324:11)
│       at Object.<anonymous> (test/logger.spec.ts:1:1)
│ -----------|---------|----------|---------|---------|-------------------
│ File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
│ -----------|---------|----------|---------|---------|-------------------
│ All files  |       0 |        0 |       0 |       0 |
│  index.ts  |       0 |        0 |       0 |       0 |
│  logger.ts |       0 |        0 |       0 |       0 | 7-156
│ -----------|---------|----------|---------|---------|-------------------
│ Test Suites: 1 failed, 1 total
│ Tests:       0 total
│ Snapshots:   0 total
│ Time:        1.612 s
│ Ran all test suites within paths .
└─ Failed in 2.4s
packages/plugins/active-directory test$ cross-env NODE_ENV=test BABEL_ENV=test jest
[2 lines collapsed]
│     Cannot find module '@verdaccio/core' from 'tests/active-directory.test.ts'
│       1 | import ActiveDirectory from 'activedirectory2';
│       2 |
│     > 3 | import { HTTP_STATUS } from '@verdaccio/core';
│         | ^
│       4 |
│       5 | import ActiveDirectoryPlugin, { NotAuthMessage } from '../src/active-directory';
│       6 | // eslint-disable-next-line jest/no-mocks-import
│       at Resolver.resolveModule (../../../node_modules/.pnpm/jest-resolve@27.3.1/node_modules/jest-resolve/build/resolver.js:324:11)
│       at Object.<anonymous> (tests/active-directory.test.ts:3:1)
└─ Running...
```

I listed only 2 of the failing packages in the output above (`logger` package included) but there there were a few others with the same exact issue related to *`Cannot find module '@verdaccio/core' from <some_other_package>`*. As mentioned, I spent some time trying to troubleshoot but I'm not intimately familiar with Verdaccio's build process/configuration or `pnpm` in general.

I have a fix for the issue from the discussion but don't want to submit the PR without working unit tests. Any guidance that can be provided would be appreciated. Thanks.

--- a/.changeset/fair-lemons-beam.md
+++ b/.changeset/fair-lemons-beam.md
@@ -1,0 +1,6 @@
+---
+'@verdaccio/logger-prettify': minor
+'@verdaccio/logger': patch
+---
+
+hydrate template placeholders in log messages when format is set to 'json'

--- a/packages/logger-prettify/src/index.ts
+++ b/packages/logger-prettify/src/index.ts
@@ -1,5 +1,7 @@
-import { printMessage } from './formatter';
+import { fillInMsgTemplate, printMessage } from './formatter';
 import { PrettyOptionsExtended } from './types';
+
+export { fillInMsgTemplate };
 
 export type PrettyFactory = (param) => string;
 
@@ -8,11 +10,11 @@ export type PrettyFactory = (param) => string;
  { messageKey: 'msg', levelFirst: true, prettyStamp: false }
  */
 
-module.exports = function prettyFactory(options: PrettyOptionsExtended): PrettyFactory {
+export default function prettyFactory(options: PrettyOptionsExtended): PrettyFactory {
   // the break line must happens in the prettify component
   const breakLike = '\n';
   return (inputData): string => {
     // FIXME: review colors by default is true
     return printMessage(inputData, options, true) + breakLike;
   };
-};
+}

--- a/packages/logger/src/logger.ts
+++ b/packages/logger/src/logger.ts
@@ -3,6 +3,7 @@ import _ from 'lodash';
 import pino from 'pino';
 
 import { warningUtils } from '@verdaccio/core';
+import prettifier, { fillInMsgTemplate } from '@verdaccio/logger-prettify';
 
 const debug = buildDebug('verdaccio:logger');
 
@@ -53,6 +54,7 @@ export function createLogger(
   // pretty logs are not allowed in production for performance reasons
   if ((format === DEFAULT_LOG_FORMAT || format !== 'json') && isProd() === false) {
     pinoConfig = Object.assign({}, pinoConfig, {
+      prettifier,
       // more info
       // https://github.com/pinojs/pino-pretty/issues/37
       prettyPrint: {
@@ -60,7 +62,23 @@ export function createLogger(
         prettyStamp: format === 'pretty-timestamped',
         ...prettyPrintOptions,
       },
-      prettifier: require('@verdaccio/logger-prettify'),
+    });
+  } else {
+    pinoConfig = Object.assign({}, pinoConfig, {
+      // more info
+      // https://github.com/pinojs/pino/blob/v7.1.0/docs/api.md#hooks-object
+      hooks: {
+        logMethod(inputArgs, method) {
+          const [templateObject, message, ...otherArgs] = inputArgs;
+          const templateVars =
+            !!templateObject && typeof templateObject === 'object'
+              ? Object.getOwnPropertyNames(templateObject)
+              : [];
+          if (!message || !templateVars.length) return method.apply(this, inputArgs);
+          const hydratedMessage = fillInMsgTemplate(message, templateObject, false);
+          return method.apply(this, [templateObject, hydratedMessage, ...otherArgs]);
+        },
+      },
     });
   }
   const logger = pino(pinoConfig, destination);

